src/Orleans.Core.Abstractions/Runtime/AsyncEnumerableRequest.cs

```diff
@@ -0,0 +1,325 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml.Schema;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Orleans.Concurrency;
+using Orleans.Invocation;
+using Orleans.Serialization.Invocation;
+
+namespace Orleans.Runtime;
+
+/// <summary>
+/// Identifies enumeration results.
+/// </summary>
+[GenerateSerializer]
+public enum EnumerationResult
+{
+    /// <summary>
+    /// This result represents a heartbeat. Issue a subsequent enumeration call to receive a new result.
+    /// </summary>
+    Heartbeat = 1,
+
+    /// <summary>
+    /// This result represents a value from the enumeration.
+    /// </summary>
+    Element = 1 << 1,
+
+    /// <summary>
+    /// This result represents a sequence of values from the enumeration.
+    /// </summary>
+    Batch = 1 << 2,
+
+    /// <summary>
+    /// This result indicates that enumeration has completed and that no further results will be produced.
+    /// </summary>
+    Completed = 1 << 3,
+
+    /// <summary>
+    /// The attempt to enumerate failed because the enumerator was not found.
+    /// </summary>
+    MissingEnumeratorError = 1 << 4,
+
+    /// <summary>
+    /// This result indicates that enumeration has completed and that no further results will be produced.
+    /// </summary>
+    CompletedWithElement = Completed | Element,
+
+    /// <summary>
+    /// This result indicates that enumeration has completed and that no further results will be produced.
+    /// </summary>
+    CompletedWithBatch = Completed | Batch,
+}
+
+/// <summary>
+/// Grain extension interface for grains which return <see cref="IAsyncEnumerable{T}"/> from grain methods.
+/// </summary>
+public interface IAsyncEnumerableGrainExtension : IGrainExtension
+{
+    /// <summary>
+    /// Invokes an <see cref="IAsyncEnumerable{T}"/> request and begins enumeration.
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    /// <param name="requestId">The request id, generated by the caller.</param>
+    /// <param name="request">The request.</param>
+    /// <returns>The result of enumeration</returns>
+    [AlwaysInterleave]
+    public ValueTask<(EnumerationResult Status, object Value)> StartEnumeration<T>(Guid requestId, [Immutable] IAsyncEnumerableRequest<T> request);
+
+    /// <summary>
+    /// Continues enumerating an <see cref="IAsyncEnumerable{T}"/> value.
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    /// <param name="requestId">The request id, generated by the caller.</param>
+    /// <returns>The result of enumeration</returns>
+    [AlwaysInterleave]
+    public ValueTask<(EnumerationResult Status, object Value)> MoveNext<T>(Guid requestId);
```

Comment:
Any reason some of these async methods don't have Async in the method name?

Reply:
It's the convention in Orleans - most of our methods don't have Async in the name. Maybe we should change that and get with the .NET norm, but the Async suffix is largely noise. Practically everything in Orleans is async because it's distributed.